File: antia_numerical_methods/f77/sub8/bspfitn.f
!	To calculate linear least squares fit to B-spline basis functions in N dimension
!	Weights are assumed to be unity
!
!	N : (input) Number of dimensions
!	NK : (input) Integer array of length N containing the number of
!		data points along each dimension
!	X : (input) Real array of length LA*N containing the coordinates
!		of points at which function values is available
!		X(I,J) contains the Ith point along Jth axis
!	F : (input) Real array of length NK(1)NK(2)...NK(N) containing the
!		function values. The dimension of F in the calling program
!		must match the size along each dimension, F(NK(1),...,NK(N))
!	K : (input) Order of B-splines required, K=4 gives cubic B-splines
!	A : (output) Real array of length LA*IV*N containing the matrix
!		U of SVD of the design matrix for fit along each axis
!	LA : (input) First dimension of arrays X and A as declared
!		in the calling program (LA .GE. 2*MAX(NK(I)))
!	V : (output) Real array of length IV*IV*N containing the matrix
!		V of SVD of the design matrix for fit along each axis
!	IV : (input) First dimension of V, XF, SIGMA in the calling program
!		IV .GE. MAX(MK(I))+K-2
!	SIGMA : (output) Real array of length IV*N containing the singular
!		values of the design matrix for fit along each axis
!	C : (output) Real array of length 2NK(1)NK(2)...NK(N) containing
!		the fitted coefficients. Note that although the number of
!		coefficients is (MK(1)+K-2)(MK(2)+K-2)...(MK(N)+K-2)
!		the rest of array is used as scratch space
!		Dimensions of array could be declared as
!		C(MK(1)+K-2, MK(2)+K-2, ..., MK(N-1)+K-2, NX)
!		where the last dimension is increased suitably to accommodate
!		the scratch space required. The last dimension NX should
!		be chosen such that the total size of array exceeds the
!		required value.
!	XF : (input) Real array of size IV*N, containing the knots
!		along each axis used for defining B-spline basis functions.
!		XF(I,J) should contain the Ith knot along Jth dimension
!	MK : (input) Integer array of length N containing the number of
!		knots for B-splines along each axis.
!	FY : (output) Real array of length NK(1)NK(2)...NK(N) containing
!		the values of fitted function at each of the tabular points
!		Dimensions of this array must match those of F.
!	WK : Real array of length 2NK(1)NK(2)...NK(N)+2LA used as scratch space
!	IWK : Integer array of length 3N used as scratch space
!	REPS : (input) Required accuracy for solution of equations using SVD
!		singular values less than REPS times maximum will be set to zero
!	RLM : (input) Parameter lambda for smoothing. If RLM.LE.0 no smoothing
!		is applied
!	IDE : (input) Order of derivative to be used for smoothing
!		This is used only when RLM>0. IDE=1 for first derivative
!		and IDE=2 for second derivative smoothing
!	CHISQ : (output) The value of Chi square at minimum
!	IER : (output) Error parameter, IER=0 implies successful execution
!		IER=609 implies that RLM>0 and IDE is not acceptable
!		No calculations are done in this case
!		Other values of IER may be set by SVD or BSPFIT
!
!	Required routines : BSPFIT, BSPLIN, BSPEVL, BSPEVN, SVD, SVDEVL
!
 
      SUBROUTINE BSPFITN(N,NK,X,F,K,A,LA,V,IV,SIGMA,C,XF,MK,FY,WK,IWK,
     1         REPS,RLM,IDE,CHISQ,IER)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION X(LA,N),A(LA*IV,*),XF(IV,N),C(*),F(*),V(IV*IV,*),
     1         SIGMA(IV,N),FY(*),WK(*),NK(N),MK(N),IWK(3*N)
 
      IF(RLM.GT.0.0.AND.(IDE.LT.1.OR.IDE.GT.2)) THEN
        IER=609
        RETURN
      ENDIF
 
      N1=4*LA+K+12
      N2=N1+LA+2
!	Set the weights to one for fits along each dimension
      DO 1000 I=0,LA
        WK(N1+I)=1.0
1000  CONTINUE
!	Calculate the SVD of matrices for fit along each dimension
      DO 2000 I=1,N
        IFLG1=1
        CALL BSPFIT(NK(I),X(1,I),F,WK(N1),K,A(1,I),LA,V(1,I),IV,
     1  SIGMA(1,I),C,XF(1,I),MK(I),FY,IFLG1,WK,REPS,RLM,IDE,CHISQ,
     2        WK(N2),IER)
        IF(IER.GT.100) RETURN
2000  CONTINUE
 
      NX=1
      DO 2200 I=1,N-1
        NX=NX*NK(I)
2200  CONTINUE
      N0=NX*NK(N)+1
      IF(RLM.GT.0.0) N0=2*NX*NK(N)+1
      NY=1
      JU=N
 
      IF(MOD(N,2).EQ.1) THEN
!	If N is odd then fit along the last dimension outside the loop
        N1=NK(N)
        IF(RLM.GT.0.0) N1=2*NK(N)
!	Set up the RHS for fit
        LJ=N1
        DO 2500 I=1,NX
          DO 2500 J=1,NK(N)
            WK(J+(I-1)*LJ)=F(I+(J-1)*NX)
            IF(RLM.GT.0.0) WK(J+NK(N)+(I-1)*LJ)=0.0
2500    CONTINUE
        M=MK(N)+K-2
        DO 2800 I=1,NX
          NI=1+(I-1)*LJ
          CALL SVDEVL(M,N1,A(1,N),V(1,N),SIGMA(1,N),LA,IV,WK(NI),WK(N0)
     1               ,REPS)
2800    CONTINUE
 
!	setup the RHS for fit along the N-1 th dimension
        NX1=NX/NK(N-1)
        NY=M
        LJ1=NK(N-1)
        IF(RLM.GT.0.0) LJ1=2*NK(N-1)
        DO 3000 I1=1,NK(N-1)
          DO 3000 I=1,NX1
            DO 3000 J=1,NY
              C(I1+(I-1)*LJ1+(J-1)*NX1*LJ1)=WK(J+(I-1)*LJ+(I1-1)*NX1*LJ)
              IF(RLM.GT.0.0) C(I1+NK(N-1)+(I-1)*LJ1+(J-1)*NX1*LJ1)=0.0
3000    CONTINUE
        NX=NX1
        JU=N-1
      ELSE
!	setup the RHS for fit along the N th dimension
        LJ=NK(N)
        IF(RLM.GT.0.0) LJ=2*NK(N)
        DO 3200 I=1,NX
          DO 3200 J=1,NK(N)
            C(J+(I-1)*LJ)=F(I+(J-1)*NX)
            IF(RLM.GT.0.0) C(J+NK(N)+(I-1)*LJ)=0.0
3200    CONTINUE
 
      ENDIF
 
!	Loop for fit along each dimension, each pass fits along 2 dimensions
      DO 5000 J1=JU,1,-2
        NUM=NX*NY
        LJ=NK(J1)
        IF(RLM.GT.0.0) LJ=2*NK(J1)
        N1=LJ
        M=MK(J1)+K-2
        DO 3400 I=1,NUM
          NI=1+(I-1)*LJ
          CALL SVDEVL(M,N1,A(1,J1),V(1,J1),SIGMA(1,J1),LA,IV,C(NI),
     1                WK(N0),REPS)
3400    CONTINUE
 
!	Set up the RHS for fit along next dimension
        NX1=NX/NK(J1-1)
        NY1=NY*M
        LJ1=NK(J1-1)
        IF(RLM.GT.0.0) LJ1=2*NK(J1-1)
        DO 3500 I1=1,NY
          DO 3500 I2=1,M
            DO 3500 I=1,NK(J1-1)
              DO 3500 J=1,NX1
                WK(I+(J-1)*LJ1+(I2-1)*LJ1*NX1+(I1-1)*NX1*LJ1*M)=
     1                      C(I2+(J-1)*LJ+(I-1)*NX1*LJ+(I1-1)*NX*LJ)
                IF(RLM.GT.0.0) WK(I+NK(J1-1)+(J-1)*LJ1+(I2-1)*LJ1*NX1+
     1                             (I1-1)*NX1*LJ1*M)=0.0
3500    CONTINUE
        NX=NX1
        NY=NY1
        NUM=NY*NX
        LJ=NK(J1-1)
        IF(RLM.GT.0.0) LJ=2*NK(J1-1)
        M=MK(J1-1)+K-2
        N1=LJ
        DO 3600 I=1,NUM
          NI=1+(I-1)*LJ
          CALL SVDEVL(M,N1,A(1,J1-1),V(1,J1-1),SIGMA(1,J1-1),LA,IV,
     1             WK(NI),WK(N0),REPS)
3600    CONTINUE
 
        IF(J1.EQ.2) THEN
!	Store the fitted coefficients in array C
          N1=NK(1)
          IF(RLM.GT.0.0) N1=2*NK(1)
          DO 3800 I=1,M
            DO 3800 J=1,NY
              C(I+(J-1)*M)=WK(I+(J-1)*LJ)
3800      CONTINUE
        ELSE
 
!	Set up the RHS for fit along next dimension
          LJ1=NK(J1-2)
          IF(RLM.GT.0.0) LJ1=2*NK(J1-2)
          M=MK(J1-1)+K-2
          NX1=NX/NK(J1-2)
          NY1=NY*M
          DO 4000 I1=1,M
            DO 4000 I2=1,NK(J1-2)
              DO 4000 I=1,NX1
                DO 4000 J=1,NY
                  C(I2+(I-1)*LJ1+(I1-1)*LJ1*NX1+(J-1)*LJ1*NX1*M)=
     1                     WK(I1+(I-1)*LJ+(I2-1)*LJ*NX1+(J-1)*LJ*NX)
                  IF(RLM.GT.0.0) C(I2+NK(J1-2)+(I-1)*LJ1+
     1                         (I1-1)*LJ1*NX1+(J-1)*LJ1*NX1*M)=0.0
4000      CONTINUE
          NX=NX1
          NY=NY1
        ENDIF
5000  CONTINUE
 
!	Calculate the Chi Square
      CHISQ=0.0
      DO 6000 I=1,N
        IWK(I)=1
6000  CONTINUE
      N0=N+1
 
!	Loop over all points
6200  IJ=IWK(1)
      WK(1)=X(IJ,1)
      ID=NK(1)
      DO 6500 I=2,N
        IJ=IJ+ID*(IWK(I)-1)
        ID=ID*NK(I)
        WK(I)=X(IWK(I),I)
6500  CONTINUE

      FY(IJ)=BSPEVN(N,MK,XF,IV,K,C,WK,WK(N0),IWK(N0),IER1)
      CHISQ=CHISQ+(F(IJ)-FY(IJ))**2
 
!	Choose the next point
      J=1
6800  IF(IWK(J).GE.NK(J)) GO TO 7000
      IWK(J)=IWK(J)+1
      GO TO 6200
 
!	If Jth dimension is exhausted go to next one
7000  IWK(J)=1
      J=J+1
      IF(J.LE.N) GO TO 6800
 
      END

File: antia_numerical_methods@Beliavsky/c/fun/bspfitn.c
/*	To calculate linear least squares fit to B-spline basis functions in N dimension
	Weights are assumed to be unity

	N : (input) Number of dimensions
	NK : (input) Integer array of length N containing the number of
		data points along each dimension
	X : (input) Array of length LA*N containing the coordinates
		of points at which function values is available
		X[j][i] contains the ith point along jth axis
	F : (input) Array of length NK[0]NK[1]...NK[N-1] containing the
		function values. The dimension of F in the calling function
		must match the size along each dimension, F[NK[N-1]]...[NK[1]]
	K : (input) Order of B-splines required, K=4 gives cubic B-splines
	A : (output) Array of length LA*IV*N containing the matrix
		U of SVD of the design matrix for fit along each axis
	LA : (input) Second dimension of array X as declared
		in the calling function (LA >= max(NK[I]))
	V : (output) Array of length IV*IV*N containing the matrix
		V of SVD of the design matrix for fit along each axis
	IV : (input) Second dimension of A, V, XF, SIGMA in the calling function
		IV >= MAX(MK[I])+K-2
	SIGMA : (output) Array of length IV*N containing the singular
		values of the design matrix for fit along each axis
	C : (output) Array of length 2NK[0]NK[1]...NK[N-1] containing
		the fitted coefficients. Note that although the number of
		coefficients is (MK[0]+K-2)(MK[1]+K-2)...(MK[N-1]+K-2)
		the rest of array is used as scratch space
		Dimensions of array could be declared as
		C[NX][MK[N-2]+K-2]...[MK[1]+K-2][MK[0]+K-2]
		where the first dimension is increased suitably to accommodate
		the scratch space required. The first dimension NX should
		be chosen such that the total size of array exceeds the
		required value.
	XF : (input) Array of size IV*N, containing the knots
		along each axis used for defining B-spline basis functions.
		XF[j][i] should contain the ith knot along jth dimension
	MK : (input) Integer array of length N containing the number of
		knots for B-splines along each axis.
	FY : (output) Array of length NK[0]NK[1]...NK[N-1] containing
		the values of fitted function at each of the tabular points
		Dimensions of this array must match those of F.
	REPS : (input) Required accuracy for solution of equations using SVD
		singular values less than REPS times maximum will be set to zero
	RLM : (input) Parameter lambda for smoothing. If RLM<=0 no smoothing
		is applied
	IDE : (input) Order of derivative to be used for smoothing
		This is used only when RLM>0. IDE=1 for first derivative
		and IDE=2 for second derivative smoothing
	CHISQ : (output) The value of Chi square at minimum

	Error status is returned by the value of the function BSPFITN.
		0 value implies successful execution
		609 implies that RLM>0 and IDE is not acceptable
		No calculations are done in this case
		Other values of may be set by SVD or BSPFIT

	Required functions : BSPFIT, BSPLIN, BSPEVL, BSPEVN, SVD, SVDEVL
*/

#include <math.h>
#include <stdlib.h>

int svdevl(int n, int m, double *u, double *v, double sigma[], int lu,
	int lv, double b[], double reps);
int bspfit(int n, double x[], double f[], double ef[], int k, double *a,
	int la, double *v, int iv, double sigma[], double c[], double xf[],
	int no, double y[], int *iflg, double reps, double rlm, int ide,
	double *chisq, double *cov);
double bspevn(int n, int nk[], double *x, int nxd, int k, double *wt,
	double x0[], int *ier);

int bspfitn(int n, int nk[], double *x, double *f, int k, double *a, int la,
	double *v, int iv, double *sigma, double *c, double *xf, int mk[],
	double *fy, double reps, double rlm, double ide, double *chisq)

{
	int i,j,i1,i2,j1,iflg,ier,nx,ny,ju,n0,n1,m,lj,nx1,lj1,ny1,num,ni,ij,id;
	double r1;
	int *iwk;
	double *wk,*cov;

	if(rlm>0.0 && (ide<1 || ide>2)) return 609;

	nx=1;
	for(i=0; i<n-1; ++i) nx=nx*nk[i];
	n0=nx*nk[n-1];
	if(rlm>0.0) n0=2*n0;

	wk=(double *) calloc((size_t) (n0), sizeof(double));
	cov=(double *) calloc((size_t) (la*la), sizeof(double));
/*	Set the weights to one for fits along each dimension */
	for(i=0; i<la; ++i) wk[i]=1.0;
 
/*	Calculate the SVD of matrices for fit along each dimension */
	for(i=0; i<n; ++i) {
		iflg=1;
		ier=bspfit(nk[i],&x[la*i],f,wk,k,&a[i*la*iv],la,&v[i*iv*iv],iv,
			&sigma[i*iv],c,&xf[i*iv],mk[i],fy,&iflg,reps,rlm,ide,chisq,cov);
		if(ier>100) {free(wk); return ier;}
	}

	ny=1;
	ju=n-1;

	if(n-2*(n/2) == 1) {
/*	If N is odd then fit along the last dimension outside the loop */
		n1=nk[n-1]; if(rlm>0.0) n1=2*nk[n-1];
/*	Set up the RHS for fit */
		lj=n1;
		for(i=0; i<nx; ++i) {
			for(j=0; j<nk[n-1]; ++j) {
				wk[j+i*lj]=f[i+j*nx];
				if(rlm>0.0) wk[j+nk[n-1]+i*lj]=0.0;
			}
		}
		m=mk[n-1]+k-2;
		for(i=0; i<nx; ++i) {
			ier=svdevl(m,n1,&a[(n-1)*la*iv],&v[(n-1)*iv*iv],&sigma[(n-1)*iv],
					la,iv,&wk[i*lj],reps);
		}
 
/*	setup the RHS for fit along the N-2 th dimension */
		nx1=nx/nk[n-2];
		ny=m;
		lj1=nk[n-2];
		if(rlm>0.0) lj1=2*lj1;
		for(i1=0; i1<nk[n-2]; ++i1) {
			for(i=0; i<nx1; ++i) {
				for(j=0; j<ny; ++j) {
					c[i1+i*lj1+j*nx1*lj1]=wk[j+i*lj+i1*nx1*lj];
					if(rlm>0.0) c[i1+nk[n-2]+i*lj1+j*nx1*lj1]=0.0;
				}
			}
		}
		nx=nx1; ju=n-2;
	}

	else {
/*	setup the RHS for fit along the N-1 th dimension */
		lj=nk[n-1];
		if(rlm>0.0) lj=2*nk[n-1];
		for(i=0; i<nx; ++i) {
			for(j=0; j<nk[n-1]; ++j) {
				c[j+i*lj]=f[i+j*nx];
				if(rlm>0.0) c[j+nk[n-1]+i*lj]=0.0;
			}
		}
	}
 
/*	Loop for fit along each dimension, each pass fits along 2 dimensions */
	for(j1=ju; j1>=0; j1 -= 2) {
		lj=nk[j1]; if(rlm>0.0) lj=2*nk[j1];
		n1=lj;
		m=mk[j1]+k-2;
		for(i=0; i<nx*ny; ++i) {
			ier=svdevl(m,n1,&a[j1*la*iv],&v[j1*iv*iv],&sigma[j1*iv],la,iv,&c[i*lj],reps);
		}

/*	Set up the RHS for fit along next dimension */
		nx1=nx/nk[j1-1];
		ny1=ny*m;
		lj1=nk[j1-1]; if(rlm>0.0) lj1=2*nk[j1-1];
		for(i1=0; i1<ny; ++i1) {
			for(i2=0; i2<m; ++i2) {
				for(i=0; i<nk[j1-1]; ++i) {
					for(j=0; j<nx1; ++j) {
						wk[i+j*lj1+i2*lj1*nx1+i1*nx1*lj1*m]=c[i2+j*lj+i*nx1*lj+i1*nx*lj];
						if(rlm>0.0) wk[i+nk[j1-1]+j*lj1+i2*lj1*nx1+i1*nx1*lj1*m]=0.0;
					}
				}
			}
		}
		nx=nx1; ny=ny1;
		lj=nk[j1-1];
		if(rlm>0.0) lj=2*nk[j1-1];
		m=mk[j1-1]+k-2;
		n1=lj;
		for(i=0; i<nx*ny; ++i) {
			ier=svdevl(m,n1,&a[(j1-1)*la*iv],&v[(j1-1)*iv*iv],&sigma[(j1-1)*iv],
					la,iv,&wk[i*lj],reps);
		}

		if(j1==1) {
/*	Store the fitted coefficients in array C */
			n1=nk[0]; if(rlm>0.0) n1=2*nk[0];
			for(i=0; i<m; ++i) {
				for(j=0; j<ny; ++j) c[i+j*m]=wk[i+j*lj];
			}
		}

		else {
/*	Set up the RHS for fit along next dimension */
			lj1=nk[j1-2]; if(rlm>0.0) lj1=2*nk[j1-2];
			m=mk[j1-1]+k-2;
			nx1=nx/nk[j1-2];
			ny1=ny*m;
			for(i1=0; i1<m; ++i1) {
				for(i2=0; i2<nk[j1-2]; ++i2) {
					for(i=0; i<nx1; ++i) {
						for(j=0; j<ny; ++j) {
							c[i2+i*lj1+i1*lj1*nx1+j*lj1*nx1*m]=wk[i1+i*lj+i2*lj*nx1+j*lj*nx];
							if(rlm>0.0) c[i2+nk[j1-2]+i*lj1+i1*lj1*nx1+j*lj1*nx1*m]=0.0;
						}
					}
				}
			}
			nx=nx1; ny=ny1;
		}
	}
 
/*	Calculate the Chi Square */
	*chisq=0.0;
	iwk=(int *) calloc((size_t) n,sizeof(int));
	for(i=0; i<n; ++i) iwk[i]=0;
 
/*	Loop over all points */
	do {
		ij=iwk[0];
		wk[0]=x[ij];
		id=nk[0];
		for(i=1; i<n; ++i) {
			ij=ij+id*iwk[i];
			id=id*nk[i];
			wk[i]=x[iwk[i]+i*la];
		}

		fy[ij]=bspevn(n,mk,xf,iv,k,c,wk,&ier);
		r1=f[ij]-fy[ij];
		*chisq=(*chisq)+r1*r1;

/*	Choose the next point */
		j=0;
		while(j <= n-1) {
			if(iwk[j]>=nk[j]-1) {
				iwk[j]=0;
				j=j+1;
			}
			else {
				iwk[j]=iwk[j]+1;
				break;
			}
		}
	} while(j<n);

	free(iwk); free(wk);
	return 0;
}
